src/database.c

```diff
@@ -481,15 +521,185 @@ _add_to_db(ProfMessage* message, char* type, const Jid* const from_jid, const Ji
+            "SELECT `from_jid`, `to_jid`, `from_resource`, `to_resource`, `message`, `timestamp`, `type`, `stanza_id`, `archive_id`, `replace_id`, `encryption`, `marked_read`"
+            "FROM `ChatLogs`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to copy V2 table.");
+        goto cleanup;
+    }
+
+    // Set replaces_db_id
+    // Note: sender check is required due to #1898 (24d0030) since previous messages might be affected.
+    query = "UPDATE `ChatLogsV2` AS A "
+            "SET `replaces_db_id` = B.`id` "
+            "FROM `ChatLogsV2` AS B "
+            "WHERE A.`replace_id` IS NOT NULL AND A.`replace_id` != '' "
+            "AND A.`replace_id` = B.`stanza_id` "
+            "AND A.`from_jid` = B.`from_jid` AND A.`to_jid` = B.`to_jid`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to add original messages to replacement messages in V2 table.");
+        goto cleanup;
+    }
+
+    // Link original messages with actual replacements
+    query = "UPDATE `ChatLogsV2` AS A "
+            "SET `replaced_by_db_id` = B.`id` "
+            "FROM `ChatLogsV2` AS B "
+            "WHERE (A.`replace_id` IS NULL OR A.`replace_id` = '') "
+            "AND A.`id` = B.`replaces_db_id` "
+            "AND A.`from_jid` = B.`from_jid`;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to link replacement messages to original messages in V2 table.");
+        goto cleanup;
+    }
+
+    // Delete old ChatLogs table
+    query = "DROP TABLE ChatLogs;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to delete old ChatLogs table.");
+        return FALSE;
+    }
+
+    // Rename ChatLogsV2 to ChatLogs
+    query = "ALTER TABLE ChatLogsV2 RENAME TO ChatLogs;";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to rename new table.");
+        return FALSE; // bad case, but unlikely to happen. Anyway, we don't want to drop it since it's the last copy
+    }
+
+    // Set new DB version (V2)
+    query = "UPDATE `DbVersion` SET `version` = 2";
+    if (SQLITE_OK != sqlite3_exec(g_chatlog_database, query, NULL, 0, &err_msg)) {
+        log_error("[DB Migration] Unable to update DB Version.");
+        return FALSE;
+    }
+
+    return TRUE;
+
+cleanup:
+    if (err_msg) {
+        log_error("SQLite error in _migrate_to_v2(): %s", err_msg);
+        sqlite3_free(err_msg);
+        err_msg = NULL;
+    } else {
+        log_error("Unknown SQLite error in _migrate_to_v2().");
+    }
+
+    query = "DROP TABLE `ChatLogsV2`;";
```

Comment:
Would it maybe make sense to turn around the logic and keep this table on failure, but drop it if it exists before starting the migration?

This would allow investigating the migration table after a failure.

Reply:
> Would it maybe make sense to turn around the logic and keep this table on failure, but drop it if it exists before starting the migration?
> 
> This would allow investigating the migration table after a failure.

In case of an error, we are going to keep the original table + migration error in the logs. We can just update the app based on this data and all that user will have to do is update app, after this original table would migrate without mistakes without need for user to interact with DB.

Reply:
Now it's `TRANSACTION`, so it's resolved, I guess.